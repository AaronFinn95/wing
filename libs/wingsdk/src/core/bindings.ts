import { IConstruct } from "constructs";
import { IInflightHost } from "../std";

const BINDINGS_SYMBOL = Symbol.for("@winglang/sdk.core.Bindings");

/**
 * Store data about what inflight hosts are bound to a construct, and
 * with what operations.
 */
export class Bindings {
  /**
   * Return the matching Bindings instance for the given construct.
   */
  public static of(construct: IConstruct): Bindings {
    let bindings = (construct as any)[BINDINGS_SYMBOL];

    if (!bindings) {
      bindings = new Bindings();
      (construct as any)[BINDINGS_SYMBOL] = bindings;
    }

    return bindings;
  }

  private readonly data = new Map<IInflightHost, Set<string>>();

  /**
   * Return a list of all inflight hosts that are bound to this construct.
   */
  public list(): Array<IInflightHost> {
    return [...this.data.keys()];
  }

  /**
   * Return a list of all operations that the given host needs to be bound to this resource with.
   */
  public get(host: IInflightHost): Array<string> {
    if (!this.data.has(host)) {
      return [];
    }

    return [...this.data.get(host)!];
  }

  /**
   * Add a binding for the given host and operations.
   */
  public add(host: IInflightHost, ...ops: string[]) {
    if (!this.data.has(host)) {
      this.data.set(host, new Set<string>());
    }

    const set = this.data.get(host)!;

    for (const op of ops) {
      set.add(op);
    }
  }

  /**
   * Return true if the given host is bound to this construct with the given operation.
   */
  public has(host: IInflightHost, op: string): boolean {
    if (!this.data.has(host)) {
      return false;
    }

    return this.data.get(host)!.has(op);
  }
}
