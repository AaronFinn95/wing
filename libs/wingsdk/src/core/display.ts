import { IConstruct } from "constructs";

const DISPLAY_SYMBOL = Symbol.for("@winglang/sdk.core.Display");

/**
 * Properties for the Display class.
 * @skipDocs
 */
export interface DisplayProps {
  /**
   * Title of the resource.
   * @default - No title.
   */
  readonly title?: string;

  /**
   * Description of the resource.
   * @default - No description.
   */
  readonly description?: string;

  /**
   * The source file or library where the resource was defined.
   * @default - No source module.
   */
  readonly sourceModule?: string;

  /**
   * Whether the resource should be hidden from the UI.
   * @default - Undefined
   */
  readonly hidden?: boolean;
}

/**
 * Information on how to display a construct in a user interface
 * such as the Wing Console.
 */
export class Display {
  /**
   * The source module for the SDK.
   */
  public static readonly SDK_SOURCE_MODULE = "@winglang/sdk";

  /**
   * Return the matching Display instance of the given construct.
   */
  public static of(construct: IConstruct): Display {
    let display = (construct as any)[DISPLAY_SYMBOL];

    if (!display) {
      display = new Display();
      (construct as any)[DISPLAY_SYMBOL] = display;
    }

    return display;
  }

  /**
   * Title of the construct.
   */
  public title?: string;

  /**
   * Description of the construct.
   */
  public description?: string;

  /**
   * The source file or library where the construct was defined.
   */
  public sourceModule?: string;

  /**
   * Whether the construct should be hidden by default in the user interface.
   */
  public hidden?: boolean;

  private constructor() {}
}
