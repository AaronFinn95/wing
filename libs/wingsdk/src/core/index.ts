export * from "./app";
export * from "./attributes";
export * from "./connections";
export * from "./dependency";
export * from "./display";
export * from "./inflight";
export * from "./tree";
