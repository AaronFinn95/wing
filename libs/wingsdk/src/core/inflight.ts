import { basename } from "path";
import { normalPath } from "../shared/misc";

/**
 * Utility class with functions about inflight clients.
 */
export class InflightClient {
  /**
   * Returns code for creating an inflight client.
   */
  public static for(
    dirname: string,
    filename: string,
    clientClass: string,
    args: string[]
  ): string {
    const inflightDir = dirname;
    const inflightFile = basename(filename).split(".")[0] + ".inflight";
    return `new (require("${normalPath(
      `${inflightDir}/${inflightFile}`
    )}")).${clientClass}(${args.join(", ")})`;
  }

  /**
   * Returns code for implementing `_toInflightType()`.
   */
  public static forType(filename: string, clientClass: string): string {
    return `require("${normalPath(filename)}").${clientClass}`;
  }

  private constructor() {}
}
