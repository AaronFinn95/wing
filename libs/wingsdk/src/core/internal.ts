import { Construct, IConstruct } from "constructs";
import { Duration } from "../std/duration";
import { IResource } from "../std/resource";
import { App } from "./app";

/**
 * Serializes `obj` into a string that can be used in an inflight scope.
 * @param obj The object to serialize
 * @param props The properties to serialize. If not specified, all properties are serialized.
 * @returns JavaScript code, not JSON!
 */
export function serializeImmutableData(scope: IConstruct, obj: any, props: string[] = []): string {
  // since typeof(null) is "object", we cover all nullity cases (undefined and null) apriori.
  if (obj == null) {
    return JSON.stringify(obj);
  }

  const tokens = App.of(scope)._tokens;
  if (tokens.isToken(obj)) {
    return tokens.lift(obj);
  }

  switch (typeof obj) {
    case "string":
    case "boolean":
    case "number":
      return JSON.stringify(obj);

    case "object":
      if (Array.isArray(obj)) {
        return `[${obj
          .map((o) => serializeImmutableData(scope, o))
          .join(",")}]`;
      }

      if (obj instanceof Duration) {
        return serializeImmutableData(scope, {
          seconds: obj.seconds,
          minutes: obj.minutes,
          hours: obj.hours,
        });
      }

      if (obj instanceof Set) {
        return `new Set(${serializeImmutableData(scope, Array.from(obj))})`;
      }

      if (obj instanceof Map) {
        return `new Map(${serializeImmutableData(scope, Array.from(obj))})`;
      }

      // if the object is a resource (i.e. has a "_toInflight" method"), we use it to serialize
      // itself.
      if (typeof (obj as IResource)._toInflight === "function") {
        return (obj as IResource)._toInflight().text;
      }

      // structs are just plain objects
      if (obj.constructor.name === "Object" || Construct.isConstruct(obj)) {
        const lines = [];
        lines.push("{");
        for (const k of props) {
          lines.push(`${k}: ${serializeImmutableData(scope, obj[k])},`);
        }
        lines.push("}");
        return lines.join("");
      }
  }

  throw new Error(
    `objects of type '${obj.constructor?.name}' cannot be referenced from an inflight scope`
  );
}
