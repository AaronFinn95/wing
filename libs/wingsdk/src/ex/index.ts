export * from "./astro";
export * from "./redis";
export * from "./table";
