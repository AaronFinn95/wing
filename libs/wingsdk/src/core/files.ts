import {
  existsSync,
  mkdirSync,
  readFileSync,
  rmSync,
  unlinkSync,
  writeFileSync,
} from "fs";
import { isAbsolute, dirname, join } from "path";
import { Construct, IConstruct } from "constructs";
import { IApp } from "./app";
import { CaptureMetadata, Code, NodeJsCode } from "./inflight";
import { Resource } from "./resource";

/**
 * Represents a file to be synthesized in the app's output directory.
 */
export abstract class FileBase extends Resource {
  /**
   * The file's relative path to the output directory.
   */
  public readonly filePath: string;

  /**
   * Indicates that generated files are not stateful. They can always be
   * regenerated from their definition.
   */
  public readonly stateful = false;

  /**
   * Defines a file
   * @param scope construct scope
   * @param id construct id
   * @param filePath relative file path
   * @param props initialization props
   */
  constructor(scope: Construct, id: string, filePath: string) {
    super(scope, id);
    this.filePath = filePath;
  }

  /**
   * Render the contents of the file and save it to the user's file system.
   */
  public save(outdir: string) {
    const data = this.render();
    const outpath = join(outdir, this.filePath);
    mkdirSync(dirname(outpath), { recursive: true });
    writeFileSync(outpath, data);
  }

  /**
   * Returns the contents of the file to save.
   */
  protected abstract render(): string;

  /**
   * @internal
   */
  public _bind(captureScope: IConstruct, metadata: CaptureMetadata): Code {
    captureScope;
    metadata;
    return NodeJsCode.fromInline(JSON.stringify(this.render()));
  }
}

/**
 * Props for `Files`.
 */
export interface FilesProps {
  /**
   * The app with files to synthesize.
   */
  readonly app: IApp;

  /**
   * The path to a state file which will track all synthesized files. If a
   * statefile is not specified, we won't be able to remove extraneous files.
   * @default - no state file
   */
  readonly stateFile?: string;
}

/**
 * Handles the synthesis of files.
 */
export class Files {
  /**
   * The path to a state file which will track all synthesized files.
   */
  public readonly stateFile?: string;
  private readonly app: IApp;
  constructor(props: FilesProps) {
    this.app = props.app;
    if (props.stateFile) {
      this.stateFile = isAbsolute(props.stateFile)
        ? props.stateFile
        : join(this.app.outdir, props.stateFile);
    }
  }

  /**
   * Synthesize the app into the output directory. The artifact produced
   * depends on what synthesizer was used.
   *
   * @param outdir The output directory, if not specified, the app's outdir will be used.
   */
  public synth(outdir?: string) {
    const oldFiles = this.readStateFile();
    const newFiles = new Set<string>();
    const isFile = (c: IConstruct): c is FileBase => c instanceof FileBase;
    if (!outdir) {
      outdir = this.app.outdir;
    }

    for (const child of this.app.node.findAll().filter(isFile)) {
      const filePath = child.filePath;
      child.save(outdir);

      oldFiles.delete(filePath);
      newFiles.add(filePath);
    }

    this.saveStateFile(newFiles);

    for (const filePath of oldFiles) {
      rmSync(join(outdir, filePath));
    }
  }

  /**
   * If a state file is defined, reads it and returns the list of files that
   * this app manages.
   */
  private readStateFile(): Set<string> {
    if (!this.stateFile) {
      return new Set();
    }

    if (!existsSync(this.stateFile)) {
      return new Set();
    }
    const files = readFileSync(this.stateFile, "utf-8").split("\n");
    return new Set(files);
  }

  /**
   * If a state file is defined, stores the list of files under management in that file.
   * @param files List of file paths (relative)
   */
  private saveStateFile(files: Set<string>) {
    if (!this.stateFile) {
      return;
    }
    if (files.size === 0) {
      if (existsSync(this.stateFile)) {
        unlinkSync(this.stateFile);
      }
      return;
    }
    writeFileSync(this.stateFile, Array.from(files).join("\n"));
  }
}
