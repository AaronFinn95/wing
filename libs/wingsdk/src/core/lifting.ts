import { IConstruct } from "constructs";
import { liftObject } from "./internal";

export class Lifting {
  /**
   * "Lifts" a value into an inflight context. If the value is a resource (i.e. has a `_toInflight`
   * method), this method will be called and the result will be returned. Otherwise, the value is
   * returned as-is.
   *
   * @param value The value to lift.
   * @returns a string representation of the value in an inflight context.
   * @internal
   */
  public static _lift(scope: IConstruct, value: any): string {
    return liftObject(scope, value);
  }
}
