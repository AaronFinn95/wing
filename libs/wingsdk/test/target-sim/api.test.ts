import { expect, test } from "vitest";
import { makeHandler } from "../../src/core/internal";
import { Api } from "../../src/target-sim";
import { SimApp } from "../../src/testing";

test("create api", async () => {
  // GIVEN
  const app = new SimApp();
  const handler = makeHandler(
    app,
    "Handler",
    `async handle(message) { console.log("Received " + message); }`
  );
  const api = new Api(app, "my_api");
  //   api.get("/hello", handler);

  const s = await app.startSimulator();

  // THEN
  await s.stop();
  expect("foo").toMatchSnapshot();
});
