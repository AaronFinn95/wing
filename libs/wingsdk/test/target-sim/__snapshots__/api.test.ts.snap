// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create api 1`] = `"foo"`;
